tests/unit/test_hash.cpp

```diff
@@ -0,0 +1,123 @@
+#include <functional>
+
+#include "hash.cpp"
+#include "util.cpp"
+
+using namespace mococrw;
+
+using testing::Eq;
+
+class HashTest : public ::testing::Test
+{
+public:
+    void SetUp() override;
+};
+
+void HashTest::SetUp()
+{
+}
+
+const auto sha256_emptyString = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+const auto sha512_emptyString = "cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e";
+const auto sha256_foo = "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae";
+const auto sha256_foobar = "c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2";
+
+TEST_F(HashTest, sha256EmptyString)
+{
+    EXPECT_THAT(utility::toHex(Hash::sha256().digest()), Eq(sha256_emptyString));
+}
+
+TEST_F(HashTest, sha512EmptyString)
+{
+    EXPECT_THAT(utility::toHex(Hash::sha512().digest()), Eq(sha512_emptyString));
+}
+
+TEST_F(HashTest, sha256SingleUpdateString)
+{
+    EXPECT_THAT(utility::toHex(Hash::sha256().update("foo").digest()), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256MultipleUpdatesString)
+{
+    EXPECT_THAT(utility::toHex(Hash::sha256().update("foo").update("bar").digest()),
+                Eq(sha256_foobar));
+}
+
+TEST_F(HashTest, sha256SingleUpdateBinaryArray)
+{
+    EXPECT_THAT(utility::toHex(Hash::sha256().update(reinterpret_cast<const uint8_t*>(&"foo"[0]), 3).digest()),
+                Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256SingleUpdateVector)
+{
+    std::string chunk = "foo";
+    std::vector<uint8_t> chunk2conv(chunk.begin(), chunk.end());
+    std::vector<uint8_t> value = Hash::sha256().update(chunk2conv).digest();
+    EXPECT_THAT(utility::toHex(value), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256StandaloneFunctionVectorVersion)
+{
+    std::string message = "foo";
+    std::vector<uint8_t> messageConv(message.begin(), message.end());
+    std::vector<uint8_t> digest = sha256(messageConv);
+    EXPECT_THAT(utility::toHex(digest), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256StandaloneFunctionBinaryVersion)
+{
+    std::string message = "foo";
+    std::vector<uint8_t> digest = sha256(reinterpret_cast<const uint8_t*>(message.c_str()), message.length());
+    EXPECT_THAT(utility::toHex(digest), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, sha256StandaloneFunctionStringVersion)
+{
+    std::vector<uint8_t> digest = sha256("foo");
+    EXPECT_THAT(utility::toHex(digest), Eq(sha256_foo));
+}
+
+TEST_F(HashTest, returnsDigestIfCalledTwice) {
+    Hash hash = Hash::sha256();
+    std::vector<uint8_t> digest1 = hash.digest();
+    std::vector<uint8_t> digest2 = hash.digest();
+    EXPECT_THAT(digest1, Eq(digest2));
+}
+
+TEST_F(HashTest, returnsDigestLength) {
+    EXPECT_THAT(Hash::sha256().getDigestLengthBytes(), Eq(32));
+    EXPECT_THAT(Hash::sha512().getDigestLengthBytes(), Eq(64));
+}
+
+TEST_F(HashTest, throwsIfCalledUpdateAfterDigest)
+{
+    std::string value;
+    EXPECT_THROW({
```

Comment:
Better confine where we expect the throw. We want it to not throw until digest and only when update() is called afterwards. This way you might mask different problems.

Reply:
done